packages/@sanity/form-builder/src/inputs/arrays/ArrayOfPrimitivesInput/ArrayOfPrimitivesInput.tsx

```diff
@@ -132,11 +171,73 @@ export class ArrayOfPrimitivesInput extends React.PureComponent<Props> {
     }
   }
 
-  handleFocusRoot = () => {
-    this.props.onFocus([])
+  handleFocusRoot = (event) => {
+    // We want to handle focus when the array input *itself* element receives
+    // focus, not when a child element receives focus, but React has decided
+    // to let focus bubble, so this workaround is needed
+    // Background: https://github.com/facebook/react/issues/6410#issuecomment-671915381
+    if (event.currentTarget === event.target && event.currentTarget === this._element) {
+      this.props.onFocus([])
+    }
   }
 
-  handleFocusItem = (item: Primitive, index: number) => {
+  getSnapshotBeforeUpdate(prevProps, prevState) {
+    const {focusPath: prevFocusPath = [], value: prevValue = []} = prevProps
+    const {focusPath = [], value = []} = this.props
+    if (prevFocusPath[0] === focusPath[0] && prevValue.length !== value.length) {
+      // the length of the array has changed, but the focus path has not, which may happen if someone inserts or removes a new item above the one currently in focus
+      const focusIndex = focusPath[0]
+
+      // the length of the array has changed, but the focus path has not, which may happen if someone inserts or removes a new item above the one currently in focus
+      const selection = window.getSelection()
+      if (!(selection.focusNode instanceof HTMLElement)) {
+        return null
+      }
+
+      const input = selection.focusNode?.querySelector('input,textarea')
+
+      return input instanceof HTMLInputElement
+        ? {
+            prevFocusedIndex: focusIndex,
+            restoreSelection: {
+              text: selection.toString(),
+              start: input.selectionStart,
+              end: input.selectionEnd,
+              value: input.value,
+            },
+          }
+        : {}
+    }
+
+    return null
+  }
+
+  componentDidUpdate(prevProps, prevState, snapshot) {
+    if (snapshot && snapshot.restoreSelection) {
```

Comment:
🧶🤌 

```suggestion
    if (snapshot?.restoreSelection) {
```

Reply:
might be better as an early return + comment too

Reply:
since it's in `componentDidUpdate`, a method that in principle can handle many different unrelated concerns I tend to prefer avoiding early returns.